path_oidc.go

```diff
@@ -383,9 +475,14 @@ func (b *jwtAuthBackend) authURL(ctx context.Context, req *logical.Request, d *f
 		return logical.ErrorResponse("role %q could not be found", roleName), nil
 	}
 
-	// If namespace will be passed around in state, and it has been provided as
+	clientNonce := d.Get("client_nonce").(string)
+	if clientNonce == "" && role.CallbackMode != callbackModeClient {
```

Comment:
That could be right.  Does it work for you if you make a new role with client callbackmode?

Reply:
Yeah, but we probably need a fallback for the empty string case for all the existing roles that exist before this is merged.

Reply:
Ok I think I have finally fixed it so an upgrade will set the default callback_mode.  @jameshartig, please test.

I also rebased against the current main.

Reply:
This works now, just tested with 1.13.2 and your latest changes here.

Reply:
@DrDaveD I spoke too soon with my last comment (now deleted). It still doesn't work with your latest changes.

The UI makes a call to `/v1/auth/oidc/oidc/auth_url` with `{"role":null,"redirect_uri":"https://$HOSTNAME/ui/vault/auth/oidc/oidc/callback"}``` but that returns a 400 error with `{"errors":["missing client_nonce"]}`.

It looks like:
<img width="731" alt="image" src="https://github.com/hashicorp/vault-plugin-auth-jwt/assets/112555/87178e98-87bc-445d-a722-b31ffb370987">

Reply:
I think the issue is that I would need to update the auth in order for it to set the callback mode to the new default but I can't log into vault to update the auth. There still needs to be a check here for empty string I think.

Reply:
I don't see how the `missing client_nonce` failure can be reached with the current code.  Please double-check that the testing was done correctly.  If you still get the error it would help if you could give me detailed step-by-step instructions of how to reproduce the problem.

There shouldn't need to be a check here because the data is read first by `pathRoleRead()`, and an empty value of role.CallbackMode encountered there gets initialized to "client" and written to storage.

Reply:
The testing is just deploying your change and then loading the index page and trying to login. I'll add some logging to try and see what's going on.

Reply:
How exactly do you deploy the change?  Does it not involve a restart of the vault process, for instance?

Reply:
It does. We deploy the binary and then restart vault via systemd.

Reply:
I don't know then why my testing showed different results, but does it work for you if you change that line to
```
if clientNonce == "" && (role.CallbackMode == callbackModeDirect || role.CallbackMode == callbackModeDevice) &&
```
?   If that works then I'll change the PR to that.

Reply:
I'm deploying this debug logging commit to help understand what's happening: https://github.com/jameshartig/vault-plugin-auth-jwt/commit/15f0e38b3981e350ec3da8de406af8610f683785

Reply:
I deployed your latest changes along with my debug logs and then went to `/` in my browser then clicked "Sign in with OIDC Provider" and got the error. When I looked at the logs the only thing that logged was:
```
2023/05/12 16:58:02 authURL loaded default with callbackMode=
```
![Recording 2023-05-12 125859](https://github.com/hashicorp/vault-plugin-auth-jwt/assets/112555/886d7b8d-bb25-49b9-9df8-d2676a4a9a94)

So it doesn't seem like `pathRoleRead` is ever being called when the `/v1/auth/oidc/oidc/auth_url` path is called.

Reply:
I can confirm that https://github.com/jameshartig/vault-plugin-auth-jwt/commit/37fefb3dc929868f592b1afb00bf5fc1756cbd89 fixes it.